### PR TITLE
Add new person

### PR DESCRIPTION
**Pull request by ngaruko**
# Description
Added scripts to create new person
Solved 'stale element' errors

### CODE REVIEW

**Comment by alxndrsn** [comment]
Please notify me when work has finished on this branch and it's ready for review.

File: tests/protractor/e2e/contacts/add-new-health-center.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+const commonElements = require('../../page-objects/common/common.po.js'),
+	contactPage = require('../../page-objects/contacts/contacts.po.js'),
+	helper = require('../../helper'),
+	utils = require('../../utils');
+
+describe('Add new health center tests : ', () => {
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/add-new-health-center.js]: You're still submitting badly formatted code - sort out your formatter or get a better text editor. We use 2 spaces for indentation, your code has 8! https://github.com/medic/medic-docs/blob/master/development/style-guide.md#indention

File: tests/protractor/e2e/contacts/add-new-health-center.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	afterEach(done => {
+		utils.resetBrowser();
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/add-new-health-center.js]: This should not be necessary and it does a browser refresh which is really slow. Why do we need to refresh?

File: tests/protractor/e2e/contacts/add-new-person.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+const commonElements = require('../../page-objects/common/common.po.js'),
+	contactPage = require('../../page-objects/contacts/contacts.po.js'),
+	helper = require('../../helper'),
+	utils = require('../../utils');
+
+describe('Add new person tests : ', () => {
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/add-new-person.js]: Same as above - indentation and unnecessary browser reset.

File: tests/protractor/e2e/contacts/add-new-person.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+	it('should add new person', () => {
+		commonElements.goToPeople();
+		expect(commonElements.isAt('contacts-list'));
+		expect(browser.getCurrentUrl()).toEqual(commonElements.getBaseUrl() + 'contacts/');
+		contactPage.addNewDistrict('BedeDistrict');
+		contactPage.completeNewPersonForm('Bede');
+		helper.waitUntilReady(element(by.css('.card-title')));
+		let name = element(by.css('span.name.ng-binding'));
+		let district = element(by.tagName('h3'));
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/add-new-person.js]: These `let`s should be `const`s.

File: tests/protractor/e2e/contacts/add-new-health-center.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		browser.actions()
+			.sendKeys('Kiwi').perform();
+		const contactName = element.all(by.css('.name')).get(1);
+		helper.waitElementToBeVisisble(contactName);
+		contactName.click();
+		element(by.css('[name="/data/health_center/external_id"]')).sendKeys('1234457');
+		element(by.css('[name="/data/health_center/notes"]')).sendKeys('some notes');
+		const submitButton = element(by.css('.btn.submit.btn-primary'));
+		submitButton.click();
+		let center = element(by.tagName('h3'));
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/add-new-health-center.js]: Should be `const`

File: tests/protractor/e2e/registration-by-sms.js [hunk 6]
```diff
@@ -1,4 +1,5 @@
-const utils = require('../utils');
+const utils = require('../utils'),
+  commonElements = require('../page-objects/common/common.po.js');
```
Comment (garethbowen) [inline tests/protractor/e2e/registration-by-sms.js]: This should line up with the `utils` above.

File: tests/protractor/e2e/report-date-filter.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 const utils = require('../utils'),
-  moment = require('moment');
+  moment = require('moment'),
+  commonElements = require('../page-objects/common/common.po.js');
```
Comment (garethbowen) [inline tests/protractor/e2e/report-date-filter.js]: These should line up with the `utils` above.

File: tests/protractor/e2e/sms-gateway.js [hunk 8]
```diff
@@ -1,4 +1,6 @@
-const utils = require('../utils');
+const utils = require('../utils'),
+  commonElements = require('../page-objects/common/common.po.js'),
+  helper=require('../helper');
```
Comment (garethbowen) [inline tests/protractor/e2e/sms-gateway.js]: Line up with the `utils` above, and add spaces around the `=` operator on line 3.

File: tests/protractor/e2e/sms-gateway.js [hunk 9]
```diff
@@ -280,12 +282,14 @@ describe('sms-gateway api', () => {
...
 
-      element(by.id('reports-tab')).click();
-
+      commonElements.goToReports();
       browser.wait(() => {
         return element(by.css('#reports-list li:first-child')).isPresent();
       }, 10000);
-      element(by.css('#reports-list li:first-child .description')).click();
+
+      let desc=element(by.css('#reports-list li:first-child .description'));
```
Comment (garethbowen) [inline tests/protractor/e2e/sms-gateway.js]: Spaces around the `=` operator.

File: tests/protractor/e2e/sms-gateway.js [hunk 10]
```diff
@@ -294,7 +298,9 @@ describe('sms-gateway api', () => {
 
       // tasks
       // State for messageId1 has been updated from pending to forwarded-to-gateway.
-      expect(element(by.css('#reports-content .details > ul .task-list .task-state .state')).getText())
+      let feedback=element(by.css('#reports-content .details > ul .task-list .task-state .state'));
+     helper.waitUntilReady(feedback);
```
Comment (garethbowen) [inline tests/protractor/e2e/sms-gateway.js]: Indentation is off by 1 space.

File: tests/protractor/e2e/submit-enketo-form.js [hunk 11]
```diff
@@ -1,6 +1,6 @@
 const utils = require('../utils'),
-      helper = require('../helper');
-
+  helper = require('../helper'),
+  commonElements = require('../page-objects/common/common.po.js');
```
Comment (garethbowen) [inline tests/protractor/e2e/submit-enketo-form.js]: Should line up with `utils` above.

File: tests/protractor/e2e/submit-enketo-form.js [hunk 12]
```diff
@@ -109,15 +109,22 @@ describe('Submit Enketo form', () => {
...
+    submitButton.click();
     browser.wait(() => {
       return element(by.css('#reports-content .details ul li:first-child p')).isPresent();
     }, 10000);
 
     browser.sleep(100); // TODO required to make the test deterministic. https://github.com/medic/medic-webapp/issues/3509
-
     // check the submitted name
-    expect(element(by.css('#reports-content .details ul li:first-child p')).getText()).toBe('Jones');
+    let detail = element(by.css('#reports-content .details ul li:first-child p'));
```
Comment (garethbowen) [inline tests/protractor/e2e/submit-enketo-form.js]: `const`. Always use `const` unless jshint tells you off, then switch to `let`.

File: tests/protractor/e2e/submit-enketo-form.js [hunk 13]
```diff
@@ -109,15 +109,22 @@ describe('Submit Enketo form', () => {
...
+    helper.waitElementToBeVisisble(detail);
+    detail.getText().then(name => {
+      expect(name).toBe('Jones');
+    }, err => {
+      console.log(err);
+    });
```
Comment (garethbowen) [inline tests/protractor/e2e/submit-enketo-form.js]: Usually we use...

```js
expect(detail.getText()).toBe('Jones');
```

Why did you change here?

File: tests/protractor/helper.js [hunk 14]
```diff
@@ -1,38 +1,58 @@
 const fs = require('fs'),
-      EC = protractor.ExpectedConditions;
+  EC = protractor.ExpectedConditions;
```
Comment (garethbowen) [inline tests/protractor/helper.js]: Should line up with `fs` above.

File: tests/protractor/helper.js [hunk 15]
```diff
@@ -1,38 +1,58 @@
...
+    browser.wait(() => {
       return element(locator).isPresent()
-        .then(presenceOfElement=> {
+        .then(presenceOfElement => {
           return !presenceOfElement;
         });
     }, 10000);
   },
+  waitElementToBeReattached: locator => {
+    //return browser.wait(() => { return retryFind(locator); }, 10000);
```
Comment (garethbowen) [inline tests/protractor/helper.js]: Remove the commented out code.

File: tests/protractor/helper.js [hunk 16]
```diff
@@ -1,38 +1,58 @@
...
+    return retryFind(locator);
+  },
 
-  waitUntilReady: elm=> {
-    return browser.wait(()=> { return elm.isPresent(); }, 10000) &&
-        browser.wait(()=> { return elm.isDisplayed(); }, 12000);
+  waitUntilReady: elm => {
+    return browser.wait(() => { return elm.isPresent(); }, 10000) &&
+      browser.wait(() => { return elm.isDisplayed(); }, 12000);
```
Comment (garethbowen) [inline tests/protractor/helper.js]: Should line up with the `browser.wait` above.

File: tests/protractor/helper.js [hunk 17]
```diff
@@ -1,38 +1,58 @@
...
       return element(locator).isPresent()
-        .then(presenceOfElement=> {
+        .then(presenceOfElement => {
           return !presenceOfElement;
         });
     }, 10000);
   },
+  waitElementToBeReattached: locator => {
+    //return browser.wait(() => { return retryFind(locator); }, 10000);
+    return retryFind(locator);
```
Comment (garethbowen) [inline tests/protractor/helper.js]: If all this function does is call `retryFind` then what's the point of it? Can't we just remove it and use `retryFind` instead?

File: tests/protractor/page-objects/common/common.po.js [hunk 18]
```diff
@@ -6,7 +6,7 @@ const messagesLink = element(by.id('messages-tab'));
 const tasksLink = element(by.id('tasks-tab'));
 const contactsLink = element(by.id('contacts-tab'));
 const analyticsLink = element(by.id('analytics-tab'));
-const reportsLink = element(by.id('reports-tab'));
+//const reportsLink = element(by.id('reports-tab'));
```
Comment (garethbowen) [inline tests/protractor/page-objects/common/common.po.js]: Remove commented out code.

File: tests/protractor/page-objects/common/common.po.js [hunk 19]
```diff
@@ -32,8 +32,10 @@ module.exports = {
   },
 
   goToReports: () => {
-    helper.waitUntilReady(reportsLink);
-    reportsLink.click();
+    let reportsLink = by.id('reports-tab');
```
Comment (garethbowen) [inline tests/protractor/page-objects/common/common.po.js]: `const`

File: tests/protractor/e2e/contacts/create-new-district.specs.js [hunk 20]
```diff
@@ -1,11 +0,0 @@
...
-      contactPage = require('../../page-objects/contacts/contacts.po.js');
-
-describe('Add new district tests : ', () => {
-  it('should open new place form', () => {
-    commonElements.goToPeople();
-    expect(commonElements.isAt('contacts-list'));
-    expect(browser.getCurrentUrl()).toEqual(commonElements.getBaseUrl() + 'contacts/');
-    contactPage.addNewDistrict('BedeDistrict', 'Bede');
-  });
-});
```
Comment (garethbowen) [inline tests/protractor/e2e/contacts/create-new-district.specs.js]: Did you delete this test or just move it?

Comment (ngaruko) [inline tests/protractor/e2e/contacts/create-new-district.specs.js]: This test is now part of add-new-person.

File: tests/protractor/e2e/submit-enketo-form.js (hunk 13 above)
Comment (ngaruko) [inline tests/protractor/e2e/submit-enketo-form.js]: This seems to fix the 'stale element' failure

**Comment by alxndrsn** [comment]
@ngaruko it looks like this was merged without fixing formatting and removing unrelated changes as mentioned by @garethbowen.
